lib/bap_llvm/llvm_value_or.hpp

```diff
@@ -0,0 +1,91 @@
+#ifndef LLVM_VALUE_OR_HPP
+#define LLVM_VALUE_OR_HPP
+
+#include <system_error>
+#include <memory>
+#include <type_traits>
+
+namespace {
+
+struct error_msg {
+    error_msg() {}
+    error_msg(const std::string &msg)    : _msg(msg) {}
+    error_msg(const error_msg& other)    : _msg(other._msg) {}
+    error_msg(const std::error_code &ec) : _msg(ec.message()) {}
+    const std::string & message() const { return _msg; }
+    std::string & message() { return _msg; }
+
+private:
+    std::string _msg;
+};
+
+// [value_or] contains either data either error message, but not both of them.
+template <typename T>
+struct value_or {
+    typedef value_or<T> self;
+    typedef T value_type;
+    typedef std::unique_ptr<T> pointer;
+
+    explicit value_or(const T &t)  : _payload(pointer(new T(t))) {}
+    explicit value_or(T &&t) : _payload(pointer(new T(std::move(t)))) {}
+    explicit value_or(T *p) : _payload(pointer(p)) {}
+    explicit value_or(const error_msg &msg)      : _payload(nullptr),  _msg(msg)   {}
+    explicit value_or(const std::error_code &ec) :  _payload(nullptr), _msg(ec)    {}
+    value_or(self &&other) : _payload(std::move(other._payload)), _msg(other._msg) {}
+
+    bool has_error() const { return _payload == nullptr; }
+    std::string message() const { return _msg.message(); }
+    const error_msg & get_error_msg() const { return _msg; }
+    error_msg & get_error_msg()             { return _msg; }
+
+    T* operator ->()            { assert_me(); return _payload.get(); }
+    const T* operator->() const { assert_me(); return _payload.get(); }
+    T & operator *()            { assert_me(); return *_payload;      }
+    const T & operator*() const { assert_me(); return *_payload;      }
+
+    // Return false if there is an error.
+    explicit operator bool() const { return !has_error(); }
+
+    pointer release() {
+        pointer p;
+        _payload.swap(p);
+        return p;
+    }
+
+    void discard(const error_msg &msg) {
+        _msg = msg;
+        _payload = nullptr;
+    }
+
+private:
+
+    value_or(const self &) { /* deprecated. use move constructor instead. */}
+    void assert_me() { assert(_payload != 0); }
+
+private:
+    pointer _payload;
+    error_msg _msg;
+};
+
+// grab error from other value_or
+template <typename D, typename S>
+value_or<D> grab_error(const value_or<S> &s) {
+    return value_or<D>(s.get_error_msg());
+}
+
+template <typename T>
+value_or<T> make_value_or(const T &t) { return value_or<T>(t); }
```

Comment:
it is better to rename these constructors to just `make_value`, e.g., `make_value(image)` makes more sense, than `make_value_or(image)` or (once you will proper rename the class name), `make_error_or(image)` Does it really mean, that an error _or_ value will be created? Depending on a flip of a coin? Nope, it will only create a value.